scipy/optimize/_zeros_py.py

```diff
@@ -1393,3 +1394,222 @@ def toms748(f, a, b, args=(), k=1,
     return _results_select(full_output, (x, function_calls, iterations, flag))
+
+
+def chandrupatla(f, x0, x1, args=(), xtol=_xtol,
+                 rtol=_rtol, maxiter=_iter, callback=None):
+
+    """
+    As written in [Sachs2015]_ which in turn is based on Chandrupatla's
+    algorithm as described in [Scherer2018]_.
+    This allows vector arguments for x0, x1, and args
+
+    Parameters
+    ----------
+
+    f : callable
+        The function whose zero is wanted. It must be a function of a
+        single variable of the form ``f(x,a,b,c...)``, where ``a,b,c...``
+        are extra arguments that can be passed in the `args` parameter.
+    x0 : float, sequence, or ndarray
+        An initial estimate of the zero that should be somewhere near the
+        actual zero. If not scalar, then `func` must be vectorized and return
+        a sequence or array of the same shape as its first argument.
+    x1 : float, optional
+        Another estimate of the zero that should be somewhere near the
+        actual zero.
+    xtol : float, optional
+        Tolerance (absolute) for termination.
+    rtol : float, optional
+        Tolerance (relative) for termination.
+    maxiter : int, optional
+        If convergence is not achieved in `maxiter` iterations, an error is
+        raised. Must be >= 0.
+    callback : callable, optional
+        An optional user-supplied function, called after each iteration.
+        Called as ``callback(xk)``, where ``xk`` is the current value of `x0`.
+
+    Returns
+    -------
+    result : OptimizeResult
+        The optimization result represented as an ``OptimizeResult`` object.
+        Important attributes are: ``x`` the solution array, ``success`` an 
+        array of Boolean flags indicating if the optimizer exited successfully,
+        ``fun``, the current function value, ``nfev``, the number of function 
+        evaluations, and ``nit``, the number of iterations. See 
+        `OptimizeResult` for a description of other attributes.
+
+    Notes
+    -----
+
+    References
+    ----------
+
+    .. [Sachs2015]
+       Sachs, Jason
+       "Ten Little Algorithms, Part 5: Quadratic Extremum Interpolation and
+       Chandrupatla's Method",
+       https://www.embeddedrelated.com/showarticle/855.php
+       
+    .. [Scherer2018]
+       Scherer, Philipp
+       *Computational physics: Simulation of classical and Quantum Systems*.
+       SPRINGER INTERNATIONAL PU, 2018. Ch. 6.1.7.3.
+       
+    See Also
+    --------
+    brentq, brenth, ridder, bisect, newton
+    root_scalar : interface to root solvers for scalar functions
+    
+    Examples
+    --------
+    >>> from scipy import optimize
+    >>> def f(x,a):
+    ...     return a-x*x
+
+    >>> k = np.arange(1,8)
+    >>> y = optimize.chandrupatla(f,0,3,args=(k,))
+    >>> y
+    [1. 1.41421356  1.73205081  2. 2.23606798  2.44948974  2.64575131]
+
+    >>> y = optimize.chandrupatla(f,0,3,args=(7.0,))
```

Comment:
I'd suggest that the first case be a purely scalar case without using `args`, and the `root` would be a scalar.

Then, the next example could be with `x0`, `x1`, and `k` mutually perpendicular arrays, so that the root would be a 3D array. This would demonstrate that broadcasting rules are followed.

Reply:
Something like this? 

```python 
x0 = np.array([1,2,1,2])
x1 = np.array([1,1,2,1])
k = np.array([1,1,2,2])

res = optimize._chandrupatla(f,x0=x0,x1=x1,
                          args=(k,),
                          maxiter=50)

res
```

produces: 

```python 
message: ['sign error' 'sign error' 'sign error' 'sign error']
       success: [False False False False]
        status: [-1 -1 -1 -1]
           fun: [       nan  1.600e-11 -4.441e-16 -4.441e-16]
             x: [       nan  1.000e+00  1.414e+00  1.414e+00]
           nit: 7
          nfev: [7 7 7 7]
 lower_bracket: [       nan  1.000e+00  1.414e+00  1.414e+00]
 upper_bracket: [       nan  1.000e+00  1.414e+00  1.414e+00]
 lower_fun_val: [       nan  1.600e-11 -4.441e-16 -4.441e-16]
 upper_fun_val: [       nan  0.000e+00  4.441e-16  4.441e-16]
```

Haven't studied mutually perpendicular arrays previously (unless I can just use orthogonal arrays)

Reply:
I don't think you'd want _all_ the cases to fail to converge. 

> Haven't studied mutually perpendicular arrays previously (unless I can just use orthogonal arrays)

I think we mean the same thing. 
One with shape (2, 1, 1), one with shape (1, 2, 1), and one with shape (1, 1, 2).

Reply:
Great, thanks, I'll look into that.